File: python/08_049.py
class TreeNode:

    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def matching_path(root1, root2):
    # Collect all paths in the first tree
    paths1 = []
    collect_paths(root1, [], paths1)
    
    # Collect all paths in the second tree
    paths2 = []
    collect_paths(root2, [], paths2)
    
    # Check for matching paths
    for path1 in paths1:
        if path1 in paths2:
            return True
    
    return False

def collect_paths(node, current_path, all_paths):
    if not node:
        return
    
    # Add current node to the path
    current_path.append(node.val)
    
    # If leaf node, add the path to all_paths
    if not node.left and not node.right:
        all_paths.append(current_path.copy())
    else:
        # Continue DFS
        collect_paths(node.left, current_path, all_paths)
        collect_paths(node.right, current_path, all_paths)
    
    # Backtrack
    current_path.pop()

File: data structures and algorithms/cpp/08_049.cpp
#include <vector>
#include <algorithm>

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
    bool matchingPath(TreeNode* root1, TreeNode* root2) {
        std::vector<std::vector<int>> paths1, paths2;
        std::vector<int> currentPath;
        
        collectPaths(root1, currentPath, paths1);
        collectPaths(root2, currentPath, paths2);
        
        for (const auto& path1 : paths1) {
            if (std::find(paths2.begin(), paths2.end(), path1) 
                    != paths2.end()) {
                return true;
            }
        }
        
        return false;
    }
    
private:
    void collectPaths(
            TreeNode* node, 
            std::vector<int>& currentPath,
            std::vector<std::vector<int>>& allPaths) {
        
        if (!node) {
            return;
        }
        
        currentPath.push_back(node->val);
        
        if (!node->left && !node->right) {
            allPaths.push_back(currentPath);
        } else {
            collectPaths(node->left, currentPath, allPaths);
            collectPaths(node->right, currentPath, allPaths);
        }
        
        currentPath.pop_back();
    }
};